File: crates/light/contracts/wire/src/v2/show_objects.rs
//! Typed v2 active-show object snapshots and output-route mutation intents.

use super::{
    dynamics::{
        DynamicActivationBoundaryProjection, DynamicPhaseSpreadModeProjection,
        DynamicRationalProjection, DynamicRunModeProjection,
    },
    events::{OutputDeliveryMode, OutputProtocol, OutputRoute, OutputRouteChange},
};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use ts_rs::TS;
use uuid::Uuid;

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct ShowObjectRecord {
    pub kind: String,
    pub id: String,
    #[ts(type = "number")]
    pub revision: u64,
    pub updated_at: String,
    #[ts(type = "unknown")]
    pub body: serde_json::Value,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub validation_error: Option<String>,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct ShowObjectCollectionSnapshot {
    pub show_id: Uuid,
    #[ts(type = "number")]
    pub show_revision: u64,
    pub kind: String,
    pub objects: Vec<ShowObjectRecord>,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct ShowObjectExactSnapshot {
    pub show_id: Uuid,
    #[ts(type = "number")]
    pub show_revision: u64,
    pub kind: String,
    pub object_id: String,
    pub object: Option<ShowObjectRecord>,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct OutputRouteActionRequest {
    pub request_id: String,
    pub action: OutputRouteAction,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum OutputRouteAction {
    Create {
        route_id: String,
        route: OutputRoute,
    },
    CreateRange {
        range_id: Uuid,
        route: OutputRoute,
        logical_universe_end: u16,
        destination_universe_end: u16,
    },
    Update {
        route_id: String,
        #[ts(type = "number")]
        expected_revision: u64,
        patch: OutputRoutePatch,
    },
    Delete {
        route_id: String,
        #[ts(type = "number")]
        expected_revision: u64,
    },
}

#[derive(Clone, Debug, Default, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct OutputRoutePatch {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub protocol: Option<OutputProtocol>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub logical_universe: Option<u16>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub destination_universe: Option<u16>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub delivery_mode: Option<OutputDeliveryMode>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub destination: Option<Option<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub enabled: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub minimum_slots: Option<u16>,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct OutputRouteActionOutcome {
    pub request_id: String,
    pub replayed: bool,
    pub changes: Vec<OutputRouteChange>,
    #[ts(type = "number")]
    pub event_sequence: u64,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct UserLayoutActionRequest {
    pub request_id: String,
    pub action: UserLayoutAction,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum UserLayoutAction {
    Update {
        #[ts(type = "number")]
        expected_revision: u64,
        patch: UserLayoutPatch,
    },
}

#[derive(Clone, Debug, Default, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct UserLayoutPatch {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(type = "unknown[] | null", optional = nullable)]
    pub desks: Option<Vec<serde_json::Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub active_desk_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(type = "unknown | null", optional = nullable)]
    pub window_settings: Option<serde_json::Value>,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct PatchLayerActionRequest {
    pub request_id: String,
    pub action: PatchLayerAction,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum PatchLayerAction {
    Save {
        #[ts(type = "number")]
        expected_revision: u64,
        layer: PatchLayerInput,
    },
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct PatchLayerInput {
    pub name: String,
    pub order: i32,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct PreloadRecordActionRequest {
    pub request_id: String,
    pub action: PreloadRecordAction,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum PreloadRecordAction {
    Preset {
        target_id: String,
        #[ts(type = "number")]
        expected_revision: u64,
        name: String,
        mode: PreloadPresetMode,
        family: PreloadPresetFamily,
    },
    Cue {
        cue_list_id: String,
        #[ts(type = "number")]
        expected_revision: u64,
        cue_number: f64,
        name: Option<String>,
    },
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
#[serde(rename_all = "snake_case")]
pub enum PreloadPresetMode {
    Merge,
    Overwrite,
    AddMissingFixtures,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize, TS)]
#[serde(rename_all = "snake_case")]
pub enum PreloadPresetFamily {
    Mixed,
    Intensity,
    Color,
    Position,
    Beam,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct ShowObjectActionOutcome {
    pub request_id: String,
    pub replayed: bool,
    pub show_id: Uuid,
    #[ts(type = "number")]
    pub show_revision: u64,
    pub object: ShowObjectRecord,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(as = "Option<f64>", optional = nullable)]
    pub event_sequence: Option<u64>,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct DynamicCreateActionRequest {
    pub request_id: String,
    /// Complete candidate definition. The server owns UUID, revision, and atomic slot conflict
    /// validation; the first lane must already be present and valid.
    #[ts(type = "unknown")]
    pub definition: serde_json::Value,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct DynamicPoolActionRequest {
    pub request_id: String,
    #[ts(type = "number")]
    pub expected_revision: u64,
    pub pool_number: u16,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct DynamicDeleteActionRequest {
    pub request_id: String,
    #[ts(type = "number")]
    pub expected_revision: u64,
}

#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
pub struct DynamicUpdateActionRequest {
    pub request_id: String,
    #[ts(type = "number")]
    pub expected_revision: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    #[ts(optional = nullable)]
    pub mutation_group: Option<String>,
    pub intent: DynamicUpdateIntent,
}

/// One deliberate editor mutation. Domain-shaped payloads remain JSON at the transport package
/// boundary and are decoded and validated into `light-dynamics` types by the server adapter.
#[derive(Clone, Debug, Deserialize, JsonSchema, PartialEq, Serialize, TS)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum DynamicUpdateIntent {
    SetName {
        name: String,
    },
    SetColor {
        color: Option<String>,
    },
    SetIcon {
        icon: Option<String>,
    },
    SetTargetBinding {
        #[ts(type = "unknown")]
        target_binding: serde_json::Value,
    },
    AddLane {
        #[ts(type = "unknown")]
        lane: serde_json::Value,
        index: Option<usize>,
    },
    ReplaceLane {
        lane_id: Uuid,
        #[ts(type = "unknown")]
        lane: serde_json::Value,
    },
    DeleteLane {
        lane_id: Uuid,
    },
    MoveLane {
        lane_id: Uuid,
        index: usize,
    },
    SetPhase {
        #[ts(type = "unknown")]
        phase: serde_json::Value,
    },
    SetPhaseMode {
        phase_mode: DynamicPhaseSpreadModeProjection,
    },
    SetSpeed {
        #[ts(type = "unknown")]
        speed: serde_json::Value,
    },
    SetOverallSpeedMultiplier {
        multiplier: DynamicRationalProjection,
    },
    SetRunMode {
        run_mode: DynamicRunModeProjection,
    },
    SetActivation {
        #[ts(type = "unknown")]
        activation: serde_json::Value,
    },
    SetActivationBoundary {
        boundary: DynamicActivationBoundaryProjection,
    },
    AddRandomGroup {
        #[ts(type = "unknown")]
        group: serde_json::Value,
    },
    ReplaceRandomGroup {
        group_id: Uuid,
        #[ts(type = "unknown")]
        group: serde_json::Value,
    },
    DeleteRandomGroup {
        group_id: Uuid,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn output_route_update_is_partial_and_tolerant() {
        let request: OutputRouteActionRequest = serde_json::from_value(serde_json::json!({
            "request_id": "route-1",
            "action": {
                "type": "update",
                "route_id": "main",
                "expected_revision": 3,
                "patch": {"enabled": false, "future_patch": true},
                "future_action": true
            },
            "future_root": true
        }))
        .unwrap();
        assert!(matches!(
            request.action,
            OutputRouteAction::Update { patch, .. } if patch.enabled == Some(false)
        ));
    }
}

File: apps/light-desktop/src/api/client/showObjects.ts
import type { PresetFamily } from "../../presetFamilies";
import { ApiRequestError } from "../ApiRequestError";
import type {
	DynamicCreateActionRequest,
	DynamicDeleteActionRequest,
	DynamicPoolActionRequest,
	DynamicSpatialPreviewRequest,
	DynamicSpatialPreviewResponse,
	DynamicUpdateActionRequest,
	DynamicUpdateIntent,
	OutputRoute,
	OutputRouteAction,
	OutputRouteActionOutcome,
	PreloadRecordAction,
	ShowObjectActionOutcome,
	ShowObjectCollectionSnapshot,
	ShowObjectExactSnapshot,
	UserLayoutPatch,
} from "../generated/light-wire";
import type { OutputRouteRangeIntent, VersionedObject } from "../types";
import { type ClientTransport, jsonRequest } from "./transport";

interface PreloadStoreInput {
	target: "preset" | "cue";
	target_id: string;
	cue_number?: string;
	name?: string;
	mode?: "merge" | "overwrite" | "add_missing_fixtures";
	family?: PresetFamily;
}

interface PatchLayerSaveInput {
	id: string;
	name: string;
	order: number;
}

export class ShowObjectsApiClient {
	constructor(private readonly transport: ClientTransport) {}

	async objects<T>(
		showId: string,
		kind: string,
	): Promise<VersionedObject<T>[]> {
		return (await this.collectionSnapshot<T>(showId, kind)).objects;
	}

	async collectionSnapshot<T>(showId: string, kind: string) {
		const snapshot = await this.transport.request<ShowObjectCollectionSnapshot>(
			`/api/v2/objects/${encodeURIComponent(kind)}`,
			{ headers: showScopeHeaders(showId) },
		);
		return {
			objects: snapshot.objects as VersionedObject<T>[],
			showRevision: snapshot.show_revision,
		};
	}

	async object<T>(showId: string, kind: string, id: string) {
		const object = await this.objectOrNull<T>(showId, kind, id);
		if (!object) throw new ApiRequestError("show object not found", 404);
		return object;
	}

	async objectOrNull<T>(showId: string, kind: string, id: string) {
		const snapshot = await this.transport.request<ShowObjectExactSnapshot>(
			v2ShowObjectPath(kind, id),
			{ headers: showScopeHeaders(showId) },
		);
		return snapshot.object as VersionedObject<T> | null;
	}

	createDynamic(
		showId: string,
		definition: import("../generated/light-wire").DynamicDefinitionProjection,
	): Promise<ShowObjectActionOutcome> {
		const request: DynamicCreateActionRequest = {
			request_id: crypto.randomUUID(),
			definition,
		};
		return this.dynamicAction("/api/v2/dynamics/create", showId, request);
	}

	moveDynamic(
		showId: string,
		id: string,
		expectedRevision: number,
		poolNumber: number,
	): Promise<ShowObjectActionOutcome> {
		const request: DynamicPoolActionRequest = {
			request_id: crypto.randomUUID(),
			expected_revision: expectedRevision,
			pool_number: poolNumber,
		};
		return this.dynamicAction(
			`/api/v2/dynamics/${encodeURIComponent(id)}/move`,
			showId,
			request,
		);
	}

	copyDynamic(
		showId: string,
		id: string,
		expectedRevision: number,
		poolNumber: number,
	): Promise<ShowObjectActionOutcome> {
		const request: DynamicPoolActionRequest = {
			request_id: crypto.randomUUID(),
			expected_revision: expectedRevision,
			pool_number: poolNumber,
		};
		return this.dynamicAction(
			`/api/v2/dynamics/${encodeURIComponent(id)}/copy`,
			showId,
			request,
		);
	}

	deleteDynamic(
		showId: string,
		id: string,
		expectedRevision: number,
	): Promise<ShowObjectActionOutcome> {
		const request: DynamicDeleteActionRequest = {
			request_id: crypto.randomUUID(),
			expected_revision: expectedRevision,
		};
		return this.dynamicAction(
			`/api/v2/dynamics/${encodeURIComponent(id)}/delete`,
			showId,
			request,
		);
	}

	updateDynamic(
		showId: string,
		id: string,
		expectedRevision: number,
		intent: DynamicUpdateIntent,
		mutationGroup?: string,
	): Promise<ShowObjectActionOutcome> {
		const request: DynamicUpdateActionRequest = {
			request_id: crypto.randomUUID(),
			expected_revision: expectedRevision,
			mutation_group: mutationGroup ?? null,
			intent,
		};
		return this.dynamicAction(
			`/api/v2/dynamics/${encodeURIComponent(id)}/update`,
			showId,
			request,
		);
	}

	previewDynamicSpatialMapping(
		showId: string,
		id: string,
		requestBody: DynamicSpatialPreviewRequest,
	): Promise<DynamicSpatialPreviewResponse> {
		const request = jsonRequest("POST", requestBody);
		return this.transport.request(
			`/api/v2/dynamics/${encodeURIComponent(id)}/spatial-preview`,
			{
				...request,
				headers: { ...request.headers, ...showScopeHeaders(showId) },
			},
		);
	}

	saveOutputRoute(
		showId: string,
		id: string,
		route: OutputRoute,
		revision: number,
	): Promise<OutputRouteActionOutcome> {
		const action: OutputRouteAction =
			revision === 0
				? { type: "create", route_id: id, route }
				: {
						type: "update",
						route_id: id,
						expected_revision: revision,
						patch: route,
					};
		return this.outputRouteAction(showId, action);
	}

	createOutputRouteRange(
		showId: string,
		intent: OutputRouteRangeIntent,
	): Promise<OutputRouteActionOutcome> {
		return this.outputRouteAction(showId, {
			type: "create_range",
			range_id: crypto.randomUUID(),
			route: {
				...intent.route,
				logical_universe: intent.logical_start,
				destination_universe: intent.destination_start,
			},
			logical_universe_end: intent.logical_end,
			destination_universe_end: intent.destination_end,
		});
	}

	deleteOutputRoute(
		showId: string,
		id: string,
		revision: number,
	): Promise<OutputRouteActionOutcome> {
		return this.outputRouteAction(showId, {
			type: "delete",
			route_id: id,
			expected_revision: revision,
		});
	}

	updateUserLayout(
		showId: string,
		sessionId: string,
		patch: {
			desks: unknown[];
			activeDeskId: string;
			windowSettings?: unknown;
		},
		revision: number,
	): Promise<ShowObjectActionOutcome> {
		const wirePatch: UserLayoutPatch = {
			desks: patch.desks,
			active_desk_id: patch.activeDeskId,
			...(patch.windowSettings === undefined
				? {}
				: { window_settings: patch.windowSettings }),
		};
		return this.showObjectAction(
			`/api/v2/user-layouts/${encodeURIComponent(sessionId)}/update`,
			showId,
			{
				type: "update",
				expected_revision: revision,
				patch: wirePatch,
			},
		);
	}

	savePatchLayer(
		showId: string,
		layer: PatchLayerSaveInput,
		revision: number,
	): Promise<ShowObjectActionOutcome> {
		return this.showObjectAction(
			`/api/v2/patch/layers/${encodeURIComponent(layer.id)}/update`,
			showId,
			{
				type: "save",
				expected_revision: revision,
				layer: { name: layer.name, order: layer.order },
			},
		);
	}

	storePreload(
		showId: string,
		input: PreloadStoreInput,
		revision: number,
	): Promise<ShowObjectActionOutcome> {
		let action: PreloadRecordAction;
		if (input.target === "preset") {
			action = {
				type: "preset",
				target_id: input.target_id,
				expected_revision: revision,
				name: input.name ?? `Preset ${input.target_id}`,
				mode: input.mode ?? "merge",
				family: preloadFamily(input.family),
			};
		} else {
			const cueNumber = input.cue_number;
			if (
				cueNumber === undefined ||
				!/^(?:0|[1-9]\d*)(?:\.(?:0|[1-9]\d*))*$/.test(cueNumber)
			) {
				throw new Error(
					"A canonical dotted cue path is required when storing to a cue",
				);
			}
			action = {
				type: "cue",
				cue_list_id: input.target_id,
				expected_revision: revision,
				cue_number: cueNumber,
				name: input.name ?? null,
			};
		}
		return this.showObjectAction("/api/v2/preload/record", showId, action);
	}

	private outputRouteAction(
		showId: string,
		action: OutputRouteAction,
	): Promise<OutputRouteActionOutcome> {
		const request = jsonRequest("POST", {
			request_id: crypto.randomUUID(),
			action,
		});
		return this.transport
			.request("/api/v2/output-routes/actions", {
				...request,
				headers: { ...request.headers, ...showScopeHeaders(showId) },
			})
			.then((outcome) => outcome as OutputRouteActionOutcome);
	}

	private dynamicAction(
		path: string,
		showId: string,
		requestBody:
			| DynamicCreateActionRequest
			| DynamicPoolActionRequest
			| DynamicDeleteActionRequest
			| DynamicUpdateActionRequest,
	): Promise<ShowObjectActionOutcome> {
		const request = jsonRequest("POST", requestBody);
		return this.transport
			.request(path, {
				...request,
				headers: { ...request.headers, ...showScopeHeaders(showId) },
			})
			.then((outcome) => outcome as ShowObjectActionOutcome);
	}

	private showObjectAction(
		path: string,
		showId: string,
		action:
			| import("../generated/light-wire").UserLayoutAction
			| import("../generated/light-wire").PatchLayerAction
			| PreloadRecordAction,
	): Promise<ShowObjectActionOutcome> {
		const request = jsonRequest("POST", {
			request_id: crypto.randomUUID(),
			action,
		});
		return this.transport
			.request(path, {
				...request,
				headers: { ...request.headers, ...showScopeHeaders(showId) },
			})
			.then((outcome) => outcome as ShowObjectActionOutcome);
	}
}

function v2ShowObjectPath(kind: string, id: string): string {
	return `/api/v2/objects/${encodeURIComponent(kind)}/${encodeURIComponent(id)}`;
}

function showScopeHeaders(showId: string): HeadersInit {
	return { "x-tosk-show": showId };
}

function preloadFamily(family: PresetFamily | undefined) {
	switch (family) {
		case "Intensity":
			return "intensity" as const;
		case "Color":
			return "color" as const;
		case "Position":
			return "position" as const;
		case "Beam":
			return "beam" as const;
		default:
			return "mixed" as const;
	}
}
